### PR TITLE
[memory leak]bpo-16322, bpo-27426: Fix time zone names encoding issues in Windows

### PR DESCRIPTION
**Pull request by denis-osipov**
Use Windows APIs to avoid wrong encoding of time zone names: [bpo-16322](https://bugs.python.org/issue16322) and [bpo-27426](https://bugs.python.org/issue27426).
Switch back to using wcsftime because:
- replace wcsftime() with strftime() to fix encoding issues doesn't work anymore ( http://bugs.python.org/issue8304#msg255133 );
- not to convert new time zone names from Windows API to char*.

https://bugs.python.org/issue16322

### CODE REVIEW

File: Modules/timemodule.c [hunk 1]
```diff
@@ -109,6 +109,20 @@ win_perf_counter(_Py_clock_info_t *info)
     }
     return PyFloat_FromDouble(diff / (double)cpu_frequency);
 }
+
+// Function to get time zone name with Windows API
+static void get_windows_zone(wchar_t *out)
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: Move "static void" on a separate line.

File: Modules/timemodule.c [hunk 2]
```diff
@@ -109,6 +109,20 @@ win_perf_counter(_Py_clock_info_t *info)
...
+{
+    TIME_ZONE_INFORMATION tzi;
+    DWORD tzid = GetTimeZoneInformation(&tzi);
+
+    if (tzid < 2) {
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: Wouldn't be better to use named constants? TIME_ZONE_ID_STANDARD, TIME_ZONE_ID_DAYLIGHT, etc.

Handle error (TIME_ZONE_ID_INVALID).

File: Modules/timemodule.c [hunk 3]
```diff
@@ -622,27 +656,27 @@ time_strftime(PyObject *self, PyObject *args)
     fmt = PyBytes_AS_STRING(format);
 #endif
 
-#if defined(MS_WINDOWS) && !defined(HAVE_WCSFTIME)
+#if defined(MS_WINDOWS) && defined(HAVE_WCSFTIME)
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: What abut the case `defined(MS_WINDOWS) && !defined(HAVE_WCSFTIME)`?

File: Modules/timemodule.c [hunk 4]
```diff
@@ -656,6 +690,45 @@ time_strftime(PyObject *self, PyObject *args)
...
+
+    // Count the number of %Z occurences
+    ins = fmt;
+    for (count = 0; tmp = wcsstr(ins, L"%Z"); ++count) {
+        ins = tmp + 2;
+    }
+
+    //Replace %Z with time zone name
+    if (count) {
+        size_t l = wcslen(fmt) + (len_zone - 2) * count + 1;
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: Check for integer overflow.
```
if (len_zone - 2 > (Py_SSIZE_MAX/sizeof(time_char) - 1 - wcslen(fmt)) / count) {
    // raise MemoryError
}
```

File: Modules/timemodule.c [hunk 5]
```diff
@@ -656,6 +690,45 @@ time_strftime(PyObject *self, PyObject *args)
...
+    const wchar_t *ins;
+    wchar_t zone[128];
+    size_t len_zone, len_copy, count;
+
+    get_windows_zone(zone);
+    len_zone = wcslen(zone);
+
+    // Count the number of %Z occurences
+    ins = fmt;
+    for (count = 0; tmp = wcsstr(ins, L"%Z"); ++count) {
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: You could use the same variable for `ins` and `tmp` here.

File: Modules/timemodule.c [hunk 6]
```diff
@@ -1222,8 +1300,17 @@ PyInit_timezone(PyObject *m) {
     PyModule_AddIntConstant(m, "altzone", timezone-3600);
 #endif
     PyModule_AddIntConstant(m, "daylight", daylight);
+    /*bpo-16322, bpo-27426: For Windows use GetTimeZoneInformation() to avoid
+    wrong encoding of time zone names.*/
+#ifdef MS_WINDOWS
+    TIME_ZONE_INFORMATION tzi;
+    GetTimeZoneInformation(&tzi);
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: Handle error (TIME_ZONE_ID_INVALID).

Is `DaylightName` initialized properly if `GetTimeZoneInformation() != TIME_ZONE_ID_DAYLIGHT`?

File: Modules/timemodule.c [hunk 7]
```diff
@@ -622,27 +656,27 @@ time_strftime(PyObject *self, PyObject *args)
     fmt = PyBytes_AS_STRING(format);
 #endif
 
-#if defined(MS_WINDOWS) && !defined(HAVE_WCSFTIME)
+#if defined(MS_WINDOWS) && defined(HAVE_WCSFTIME)
     /* check that the format string contains only valid directives */
-    for (outbuf = strchr(fmt, '%');
+    for (outbuf = wcschr(fmt, L'%');
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: Loops for Windows, AIX and Solaris are too similar. Can they be merged?

File: Modules/timemodule.c [hunk 8]
```diff
@@ -656,6 +690,45 @@ time_strftime(PyObject *self, PyObject *args)
...
+    }
+
+    //Replace %Z with time zone name
+    if (count) {
+        size_t l = wcslen(fmt) + (len_zone - 2) * count + 1;
+        tmp = result = (time_char *)PyMem_Malloc(l * sizeof(time_char));
+        while (count--) {
+            ins = wcsstr(fmt, L"%Z");
+            len_copy = ins - fmt;
+            if (wcsncmp(ins - 1, L"%", 1) == 0) {
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c]: If the format starts with a "%Z", `ins == fmt` and `ins - 1` is outside of a buffer.

If "%Z" follows a double "%" ("%%%Z"), it is copied as is.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Modules/timemodule.c (hunk 1 above)
Comment (denis-osipov) [reply Modules/timemodule.c]: Done.

File: Modules/timemodule.c (hunk 3 above)
Comment (denis-osipov) [reply Modules/timemodule.c]: @serhiy-storchaka If I'm not mistaken, this case is possible only when we undefine HAVE_WCSFTIME somewhere. Do you mean we should have special code for this case?
Now I just returned this block to state before switching from wcsftime to strftime.

File: Modules/timemodule.c (hunk 6 above)
Comment (denis-osipov) [reply Modules/timemodule.c]: Yes, I think so. `TIME_ZONE_INFORMATION` structure get all members if `GetTimeZoneInformation()` succeeds. On my system `GetTimeZoneInformation() == TIME_ZONE_ID_UNKNOWN`, i.e. daylight saving time is not used, but both names initialized properly. However, `DaylightName` and `StandardName` can be empty.

**Comment by denis-osipov** [comment]
I didn't expect the Spanish Inquisition!

**Comment by denis-osipov** [comment]
Oops... Fix of memory leak needed.
```
./python.bat -bb -E -Wd -m test test_strftime -r -uall -j0 -R :
...
test_strftime leaked [1, 2, 1, 1] memory blocks, sum=5
...
```

**Pull request closed without merging**